Library/Homebrew/dev-cmd/linkage.rb

```diff
@@ -68,11 +68,16 @@ def check_dylibs
 
       begin
         f = Formulary.from_rack(keg.rack)
-        @undeclared_deps = @brewed_dylibs.keys - f.deps.map(&:name)
+        f.build = Tab.for_keg(keg)
+        filter_out = proc do |dep|
+          dep.build? || (dep.optional? && !dep.option_names.any? { |n| f.build.with?(n) })
+        end
+        declared_deps = f.deps.reject { |dep| filter_out.call(dep) }.map(&:name) +
+                        f.requirements.reject { |req| filter_out.call(req) }.map(&:default_formula).compact
```

Comment:
The logic are different. Noted, for `brew deps/uses`, they will ignore local installation completely as they only query on deps graph.